tests/func/test_api.py

```diff
@@ -143,3 +149,46 @@ def _set_remote_url_and_commit(repo, remote_url):
     rconfig.modify("upstream", "url", remote_url)
     repo.scm.add([repo.config.config_file])
     repo.scm.commit("modify remote")
+
+
+def test_open_scm_controlled(dvc_repo, repo_dir):
+    stage, = dvc_repo.add(repo_dir.FOO)
+
+    stage_content = open(stage.path, "r").read()
+    with api.open(stage.path) as fd:
+        assert fd.read() == stage_content
+
+
+def test_open_not_cached(dvc_repo, repo_dir):
+    metric_file = "metric.txt"
+    code_file = "metric_writing_code.py"
+    metric_content = "0.6"
+    metric_code = 'open("{}", "w").write("{}")'.format(
+        metric_file, metric_content
+    )
+    repo_dir.create(code_file, metric_code)
+
+    dvc_repo.run(
+        metrics_no_cache=[metric_file], cmd="python {}".format(code_file)
+    )
+
+    with api.open(metric_file) as fd:
+        assert fd.read() == metric_content
+
+
+def test_open_missing_not_cached(dvc_repo, repo_dir):
+    metric_file = "metric.txt"
+    code_file = "metric_writing_code.py"
+    metric_content = "0.6"
+    metric_code = 'open("{}", "w").write("{}")'.format(
+        metric_file, metric_content
+    )
+    repo_dir.create(code_file, metric_code)
+
+    dvc_repo.run(
+        metrics_no_cache=[metric_file], cmd="python {}".format(code_file)
+    )
+    os.remove(metric_file)
+
+    with pytest.raises(NotCachedOutputFileMissingError):
+        api.read(metric_file)
```

Comment:
These two tests are almost the same. I would say simply move:
```python
    os.remove(metric_file)
    with pytest.raises(NotCachedOutputFileMissingError):
        api.read(metric_file)
```
to the end of the first one and remove the second test. 

Also, why do you create new code file, while we already have `code.py`?

Reply:
New code is required due to issue regarding original code described in #2574